### ISSUE TITLE
slow pylint performance with a hot spot in astroid

### ISSUE THREAD

**Issue by robin-wayve** [PyCQA/astroid#1115]
tl;dr I have been trying to figure out why `pylint` is slow for us and profiling points to `astroid`. I'd appreciate some pointers on how I could get a clearer signal on what is going on.

### Versions

```
astroid 2.6.5
pylint 2.9.6
python 3.7.9
```

### Description

We run `pylint` against individual packages/modules in our codebase and the performance is highly variable. I've been trying to figure out why some runs are slow with `yappi`, similar to how it was done in #610.

One such module of ours has about 1000 lines of code and has become infeasibly slow on the versions listed above (I gave up after waiting 1 hour with the profiler on). I managed to get it to complete a profile in ~30 minutes by limiting the run to a single 200 line file. It does complete in under 5 minutes on our current versions: `pylint` 2.7.4 with `astroid` 2.5.2 -- I haven't captured a profile on these versions yet, but I could if that would be helpful.

I'm not able to share the code since it is proprietary, and the profile may contain sensitive information. I may attempt to scrub the profile and share the dot graph if that would be of interest.

### Profling result

The hottest node from the `callgrind` was this by quite some margin:
```
Name.infer astroid/node_classes.py:325 57.62% (16.12%) 82185683×
```
I believe that refers to this code: https://github.com/PyCQA/astroid/blob/v2.6.5/astroid/node_classes.py#L325-L379

This node has quite a lot of edges making it hard to see exactly what is going on but I hope this an interesting enough starting point for us to figure out what the cause of the bottleneck could be?

**Comment by robin-wayve** [comment]
There's quite a few other details I'm not sure whether to include: 

- to get a useful profile, this was done with `--jobs=1`: it does run faster with `--jobs=0` but it's still quite slow.
- I believe our `rcfile` is relatively boring apart from disabling quite a lot of pylint's checkers
- Other options we're using
```
--persistent=n --unsafe-load-any-extension=y
```

**Comment by Pierre-Sassoulas** [comment]
Thank you for opening the issue. It makes sense that the inference would be what is costing a lot on astroid. We'd need to dig deeper and analyse what inferences exactly are taking more time than they should and probably in what pylint checker.

**Comment by Pierre-Sassoulas** [comment]
Coule you try to single out deep recursion call, if possible ?

**Comment by robin-wayve** [comment]
I tidied up the graph, hope this will help:
![pylint296_scrubbed](https://user-images.githubusercontent.com/70587124/127857358-b3b4fc6e-006c-439c-ae40-286a7fd6800b.png)

**Comment by Pierre-Sassoulas** [comment]
Looking at your graph it looks like ClassDef.ancestors is taking quite a lot of time. Is there a lot of inheritance in your code ? Maybe this is something what we needs to cache.

**Comment by robin-wayve** [comment]
> Looking at your graph it looks like ClassDef.ancestors is taking quite a lot of time. Is there a lot of inheritance in your code ? Maybe this is something what we needs to cache.

Yes, and the inherited classes are also using Templates / Generics which I think have read is another known slow thing?

**Comment by Pierre-Sassoulas** [comment]
I changed the name of the issue even if this might not be the only problem with the slowness of astroid on your code. This is the one issue that jump to the eye and we already had trouble with this before. Thank you for providing a graphs this is really helpful.

**Comment by nickdrozd** [comment]
That red box for `Name.infer` is an opportunity for improvement.

**Comment by robin-wayve** [comment]
I was also going to ask: is there a way I could figure out the minimum set of pylint checkers to disable in this context to keep my runs fast in the meantime?

**Comment by Pierre-Sassoulas** [comment]
There is no easy way right now. We can't profile each checker independently (and they are often intertwined)  There is an issue opened in pylint to provide the tools for that: https://github.com/PyCQA/pylint/issues/4067.

**Comment by dwo** [comment]
@Pierre-Sassoulas I started a rough sketch of caching `ClassDef.ancestors` in #1120 which still fails two tests, but is this roughly the right idea?

**Comment by Pierre-Sassoulas** [comment]
Yeah, I think it's smoothly the right idea 😄

**Comment by rgant** [comment]
Bumping with current data — this still reproduces five years on.

I profiled `pylint --jobs=1` against a single ~250-line file in a typical FastAPI / SQLAlchemy / Pydantic project (~21k LoC total) on pylint 4.0.5 / astroid 4.0.4 / Python 3.14.4 (darwin/arm64, GIL). Total wall: 5.68s.

```
function (file)                            thread                      total       prim    cumtime   %wall
ancestors (scoped_nodes.py)                astroid#1115               332387      57689     2.765s   48.6%
infer (node_ng.py)                         general inference          327110       6541     3.797s   66.8%
ast_from_module_name (manager.py)          astroid#1145                 6844       1022     3.592s   63.2%
_visit (transforms.py)                     brain transform visit      296630        220     2.789s   49.0%
file_build (builder.py)                    AST build cost                113         14     3.601s   63.3%
```

`ClassDef.ancestors` accounts for **48.6% of single-file wall time** (2.77s of 5.68s), with a ~5.7× recursion factor (332k total calls / 57k primitive). Same MRO-walk-recomputed-across-inference-paths shape as the 2021 report. The original reporter couldn't share their codebase; for context, the project that produced this run is stack-typical 2026 Python — SQLAlchemy 2.x ORM models, FastAPI dependencies, Pydantic models — template/generic-heavy but no metaclass acrobatics or exotic patterns.

The two big AST-build rows (`file_build` and `ast_from_module_name`, ~63% of wall apiece) are a different concern — they're the "load every transitively-imported third-party package into astroid" cost, which is #1145 territory rather than this issue. Mentioning them only because they appear in the same profile and a reader might wonder; not advocating for conflating the threads.

If it'd be useful for collecting comparable data points across other reporters' projects, I've attached `profile_pylint.py` — a self-contained PEP 723 script that reproduces the table for any target. With [`uv`](https://docs.astral.sh/uv/) installed, it's one command:

```bash
uv run profile_pylint.py path/to/your/code
```

[profile_pylint.py](https://github.com/user-attachments/files/27487822/profile_pylint.py)
[Gist copy](https://gist.github.com/rgant/a26bc387ed3fdd130b6111a297fa2f93)

The pinned-hotspot table is the part worth quoting back here — same row labels regardless of which other code paths dominate a given project, so cross-reporter comparisons stay apples-to-apples.

Happy to run it against other shapes of project (pure-data, framework-heavy, plugin-rich) if more samples would help.

**Comment by Pierre-Sassoulas** [comment]
Thank you for the analysis @rgant ! I made a patch for this, would you mind seeing how it fares in your benchmark, please ? https://github.com/pylint-dev/astroid/commit/542e40772163cc86f92b6a89762449770b41e8e1

**Comment by rgant** [comment]
@Pierre-Sassoulas Ran the patch (`542e4077`) against the same target as my previous comment, n=4 interleaved.

Same project, slightly updated ~250-line SQLAlchemy 2.x model file in a ~21k-LoC FastAPI/SQLAlchemy/Pydantic project (the file picked up a small column add in the interim, so the absolute numbers move ~2% from my last comment). pylint 4.0.5 / astroid 4.0.4 → 4.2.0b3 (your branch) / Python 3.14.4 (darwin/arm64, GIL).

```
                                                            total       prim    cumtime   %wall
baseline (astroid 4.0.4):
  ancestors (scoped_nodes.py)        astroid#1115         342,438     52,868     2.700s   51.3%
patched  (astroid 4.2.0b3 @ 542e4077):
  ancestors (scoped_nodes.py)        astroid#1115         197,502    122,236     1.806s   37.9%
```

Wall: **5.26s ± 0.06 → 4.77s ± 0.10 (-9.3%)**. `ancestors` cumtime: **-33%**. Total calls: **-42%**.

The primitive-count jump is a reporting artifact, not a regression: `ancestors(recurs=True)` is now a thin cache lookup (each call primitive), while the recursion moved into `_walk_ancestors`. Recursion factor on `ancestors` dropped from 5.7× to 1.6×.

Your commit message reports -4.0% on `pandas/core/frame.py` (21.34s → 20.48s); the gain expanding to -9.3% here is consistent with deeper MROs from SQLAlchemy 2.x `DeclarativeBase` vs pandas' flatter `NDFrame`/`DataFrame` hierarchy. LGTM, excited to see this land!

**Comment by Pierre-Sassoulas** [comment]
Sounds good ! I tried some other approachs since in f6a391bc397d7ca18cff026181a94b8c76801b26, if you don't mind confirming again.

**Comment by rgant** [comment]
@Pierre-Sassoulas confirmed against current branch HEAD `f6a391b`. The branch now carries four commits: `542e40`, `ea5b0d4`, `e89e3e7`, `f6a391b`. Same target: (now) ~351-line SQLAlchemy 2.x model file in a ~21k-LoC FastAPI/SQLAlchemy/Pydantic project. pylint 4.0.5 / Python 3.14.4 (darwin/arm64, GIL). n=8 interleaved, Latin-square rotation between arms.

```
                                                            total       prim    cumtime   %wall
baseline (astroid 4.0.4):
  ancestors (scoped_nodes.py)        astroid#1115         342,444     52,872     2.626s   65.3%
patched  (astroid 4.2.0b3 @ f6a391b):
  ancestors (scoped_nodes.py)        astroid#1115         100,178     55,493     1.609s   48.5%
```

Wall: **4.03s ± 0.07 → 3.32s ± 0.11 (-17.6% ± 1.2%)** for the full branch (all four commits at `f6a391b`) — versus -9.3% for `542e40` alone in my prior comment, so the three follow-up commits compound to ~+8pp on this workload. `ancestors` cumtime **-39%**, total calls **-71%** (342k → 100k; recursion factor 6.5× → 1.8×). `infer` total calls drop -56% (341k → 151k); `_visit` cumtime moves -18% without changing call counts — the shape `e89e3e72`'s single-`Call` dispatcher would produce.
